src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscriptionFiltered.cs

```diff
@@ -39,104 +49,445 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ResolvedEvent> {
+				private readonly Guid _subscriptionId;
+				private readonly IPublisher _bus;
+				private readonly bool _resolveLinks;
+				private readonly IEventFilter _eventFilter;
+				private readonly IPrincipal _user;
+				private readonly CancellationTokenSource _disposedTokenSource;
+				private readonly ConcurrentQueueWrapper<ResolvedEvent> _buffer;
+				private readonly CancellationTokenRegistration _tokenRegistration;
+				private Position _nextPosition;
+				private ResolvedEvent _current;
+
+				public ResolvedEvent Current => _current;
+
+				public CatchupAllSubscription(Guid subscriptionId,
+					IPublisher bus,
+					Position position,
+					bool resolveLinks,
+					IEventFilter eventFilter,
+					IPrincipal user,
+					IReadIndex readIndex,
+					CancellationToken cancellationToken) {
+					if (bus == null) {
+						throw new ArgumentNullException(nameof(bus));
+					}
+
+					if (eventFilter == null) {
+						throw new ArgumentNullException(nameof(eventFilter));
+					}
+
+					if (readIndex == null) {
+						throw new ArgumentNullException(nameof(readIndex));
+					}
+
+					_subscriptionId = subscriptionId;
+					_bus = bus;
+					_nextPosition = position == Position.End
+						? Position.FromInt64(readIndex.LastIndexedPosition, readIndex.LastIndexedPosition)
+						: position;
+					_resolveLinks = resolveLinks;
+					_eventFilter = eventFilter;
+					_user = user;
+					_disposedTokenSource = new CancellationTokenSource();
+					_buffer = new ConcurrentQueueWrapper<ResolvedEvent>();
+					_tokenRegistration = cancellationToken.Register(_disposedTokenSource.Dispose);
+					Log.Info("Catch-up subscription {subscriptionId} to $all:{eventFilter} running...",
+						_subscriptionId,
+						_eventFilter);
 				}
 
-				if (_disposedTokenSource.IsCancellationRequested) {
-					return false;
+				public ValueTask DisposeAsync() {
+					_disposedTokenSource.Dispose();
+					_tokenRegistration.Dispose();
+					return default;
 				}
 
-				try {
-					await Task.Delay(100, _disposedTokenSource.Token).ConfigureAwait(false);
-				} catch (ObjectDisposedException) {
-					return false;
+				public async ValueTask<bool> MoveNextAsync() {
+					ReadLoop:
+					if (_disposedTokenSource.IsCancellationRequested) {
+						return false;
+					}
+
+					if (_buffer.TryDequeue(out var current)) {
+						_current = current;
+						return true;
+					}
+
+					var correlationId = Guid.NewGuid();
+
+					var readNextSource = new TaskCompletionSource<bool>();
+
+					var (commitPosition, preparePosition) = _nextPosition.ToInt64();
+
+					Log.Trace(
+						"Catch-up subscription {subscriptionId} to $all:{eventFilter} reading next page starting from {nextPosition}.",
+						_subscriptionId, _eventFilter, _nextPosition);
+
+					_bus.Publish(new ClientMessage.FilteredReadAllEventsForward(
+						correlationId, correlationId, new CallbackEnvelope(OnMessage), commitPosition, preparePosition,
+						32, _resolveLinks, false, 32, default, _eventFilter, _user));
```

Comment:
See AllSubscription, `MaxCount` should be configurable.
`MaxSearchWindow` should possibly also be configurable.